UI Components: App developer documentation (#6457)
nautobot/nautobot: nautobot/docs/development/core/ui-component-framework.md

    footer_content_template_path=None,  # Footer content template
    template_path="components/panel/panel.html",  # Overall panel template
    body_wrapper_template_path="components/panel/body_wrapper_generic.html"  # Body wrapper template
)
```

### 1. KeyValueTablePanel
Displays custom key-value pairs with optional transformations.

```python
KeyValueTablePanel(
    # Data Configuration
    data={                      # Dictionary of key-value pairs to display
        "speed": "1000000",
        "notes": "**Important**"
    },
    
    hide_if_unset=(),          # Keys to hide if value is falsey
                               # Example: ["optional_field", "notes"]
    
    value_transforms={          # Transform functions for specific keys
        "speed": [             # List of functions to apply in order
            humanize_speed,    # Convert 1000000 to "1 Gbps"
            placeholder        # Show placeholder if empty
        ],
        "notes": [
            render_markdown,   # Convert markdown to HTML
            placeholder       # Show placeholder if empty
        ]
    },       # Order within section
)

> ```

Reviewer: We might need two "examples" for each panel class - one just showing the parameters and their default values, and another showing an example of real usage. I think mixing the two in a single example is confusing.

### 2. ObjectFieldsPanel
Displays object attributes in a key-value table format. This panel inherits from `KeyValueTablePanel`.

```python
ObjectFieldsPanel(
    # Field Selection

>     fields="__all__",           # List of field names to display or "__all__" for all fields

Reviewer: Let's document the fields that are implicitly *not* included in `__all__` (i.e. `created`, `last_updated`, `comments`, `tags`) and explain why and how they are automatically handled instead.

                ],
                section=SectionChoices.RIGHT_HALF,
                weight=100,
            ),
            GroupedKeyValueTablePanel(
                body_id="custom-fields",
                data={
                    "Network": {
                        "VLAN": "100",
                        "IP Range": "192.168.1.0/24"
                    }
                },
                section=SectionChoices.FULL_WIDTH,
                weight=300,
            ),
            ObjectTextPanel(
                object_field="description",
                render_as=BaseTextPanel.RenderOptions.MARKDOWN,
                section=SectionChoices.FULL_WIDTH,
                weight=400,
            )
        )
    )
```

//TODO: (include image) <img src="/api/placeholder/800/600" alt="Complete Example" />
*Complete example showing all panel types in action*

## Best Practices

### 1. Panel Organization
- Use consistent weights across your application
- Group related information in adjacent panels
- Consider mobile viewports when choosing sections

### 2. Performance
- Be specific with field selections
- Use appropriate indexes for StatsPanel queries

> - Consider lazy loading for heavy content

Reviewer: This one could be a whole topic unto itself... it's not something we currently have any built-in functionality for though?

### Why Use the UI Framework?

- ✨ **Reduced Development Time**: Eliminate boilerplate HTML/CSS
- 🎯 **Consistency**: Automatic adherence to Nautobot's design patterns
- 🔄 **Reusable Components**: Compose views from pre-built panels
- 🛠 **Extensible**: Easy to customize and extend

## Getting Started

### Basic Setup

1. Create a view that inherits from `NautobotUIViewSet`:

```python
from nautobot.apps import views
from nautobot.apps.ui import ObjectDetailContent, SectionChoices, ObjectFieldsPanel

class ExampleUIViewSet(views.NautobotUIViewSet):
    queryset = Example.objects.all()
    ...

    object_detail_content = ObjectDetailContent(
        panels=[
            ObjectFieldsPanel(
                section=SectionChoices.LEFT_HALF,
                weight=100,
                fields="__all__"
            )
        ]
    )
```

<!-- markdownlint-disable no-inline-html -->
<div class="grid cards" style="width: 300px;" markdown>

> - ![UI Framework Example](../../media/development/core/ui-component-framework/ui-framework-example.png){ .on-glb }

Reviewer 2: We have one in `Panel Examples` section. But I moved this one to `Introduction` as "more general" overview.

[Code reference](../../code-reference/nautobot/apps/ui.md#nautobot.apps.ui.Panel)

#### Panel Examples

```python
from nautobot.apps.ui import Panel, SectionChoices

Panel(
    label="Panel Header",
    section=SectionChoices.FULL_WIDTH,
    weight=100,
)
```

```python
from nautobot.apps.ui import Panel, SectionChoices

Panel(
    label="Optional Params Included",
    section=SectionChoices.RIGHT_HALF,
    weight=200,
    body_content_template_path="path/to/template/body_content_template.html",
    header_extra_content_template_path="path/to/template/header_extra_content_template.html",
    footer_content_template_path="path/to/template/footer_content_template.html",
    template_path="path/to/template/template.html",
    body_wrapper_template_path="path/to/template/body_wrapper_template.html",
)
```

### 1. ObjectFieldsPanel

`ObjectFieldsPanel` is designed to automatically render object attributes in a table format. It's particularly useful for displaying model instances or any object with defined attributes. This panel inherits from `KeyValueTablePanel`.

[Code reference](../../code-reference/nautobot/apps/ui.md#nautobot.apps.ui.ObjectFieldsPanel)

NOTE:

>     When `fields="__all__"`, the panel automatically excludes: ManyToMany fields, Reverse relations, Hidden fields and Special fields (`id`, `created`, `last_updated`, `comments`, `tags`). `comments` and `tags` are automatically added as standalone panels.

Reviewer: ```suggestion
    When `fields="__all__"`, the panel automatically excludes:
    
    - ManyToMany fields and reverse relations (these should be displayed via `ObjectsTablePanel` if desired)
    - hidden fields (name starting with `_` or explicitly declared as `hidden=True`)
    - `id`, `created`, `last_updated` (these are automatically displayed elsewhere in the standard view template)
    - `comments`, `tags` (these are automatically added as standalone panels)
```

#### ObjectFieldsPanel Examples

```python
from nautobot.apps.ui import ObjectFieldsPanel, SectionChoices

ObjectFieldsPanel(
    label="Object Fields Panel",
    section=SectionChoices.LEFT_HALF,

>     context_object_key="obj",

Reviewer: ```suggestion
    weight=100,
```

)
```

```python
from nautobot.apps.ui import ObjectFieldsPanel, SectionChoices

ObjectFieldsPanel(
    label="Object Fields Panel",
    fields=["name", "status", "type", "notes"],
    section=SectionChoices.LEFT_HALF,
    context_object_key="obj",
    ignore_nonexistent_fields=True,
    weight=200,
    value_transforms={
        "status": [status_to_badge],
        "type": [type_to_link]
    },
)
```

<!-- markdownlint-disable no-inline-html -->
<div class="grid cards" style="width: 300px;" markdown>

> - ![ObjectFieldsPanel Example](../../media/development/core/ui-component-framework/object-fields-panel-example.png){ .on-glb }

Reviewer: Would be nice in general if these screenshots directly corresponded to the above examples - but I know that will be more work to get there.

</div>
<!-- markdownlint-enable no-inline-html -->

### 2. KeyValueTablePanel

> `KeyValueTablePanel` is a Panel component that displays data in a two-column table format, commonly used in object detail views. It extends the base `Panel` class and provides additional functionality for data display and transformation.

Reviewer: ```suggestion
`KeyValueTablePanel` is a Panel component that displays data in a two-column table format, commonly used in object detail views. It extends the base `Panel` class and provides additional functionality for data display and transformation.

!!! tip
    For displaying the attributes of a specific data model instance, you'll more commonly want to use the `ObjectFieldsPanel` subclass rather than using a `KeyValueTablePanel` directly.
```

[Code reference](../../code-reference/nautobot/apps/ui.md#nautobot.apps.ui.KeyValueTablePanel)

#### KeyValueTablePanel Examples

```python
from nautobot.apps.ui import KeyValueTablePanel

KeyValueTablePanel(
    data={
        "speed": "1000000",
        "notes": "**Important**"
    },
)
```

```python
from nautobot.apps.ui import KeyValueTablePanel

KeyValueTablePanel(
    weight=100,
    data={
        "speed": "1000000",
        "notes": "**Important**",
    },
    hide_if_unset=(),
    value_transforms={
        "speed": [             # List of functions to apply in order
            humanize_speed,    # Convert 1000000 to "1 Gbps"
            placeholder        # Show placeholder if empty
        ],
        "notes": [
            render_markdown,   # Convert markdown to HTML
            placeholder       # Show placeholder if empty
        ]
    },
)
```

### 3. GroupedKeyValueTablePanel

> `GroupedKeyValueTablePanel` is a specialized version of `KeyValueTablePanel` that organizes data into collapsible accordion groups. It's particularly useful for displaying hierarchical key-value data or grouped custom fields.

Reviewer: ```suggestion
`GroupedKeyValueTablePanel` is a specialized version of `KeyValueTablePanel` that organizes data into collapsible accordion groups. This is used in the standard template for displaying grouped custom/computed fields when present.
```

[Code reference](../../code-reference/nautobot/apps/ui.md#nautobot.apps.ui.StatsPanel)

#### StatsPanel Examples

```python
from nautobot.apps.ui import StatsPanel, SectionChoices

StatsPanel(
    filter_name="location",
    related_models=[           # Models to show statistics for
        Device,               # Direct model reference
                             # Will count all devices related to this object

        (Circuit, "circuit_terminations__location__in"),
                             # Tuple of (Model, query_string)
                             # For complex relationships

        (VirtualMachine, "cluster__location__in")
                             # Another complex relationship example
    ],
    label="Statistics",
    section=SectionChoices.RIGHT_HALF,
    weight=700,
)
```

<!-- markdownlint-disable no-inline-html -->
<div class="grid cards" style="width: 300px;" markdown>
- ![StatsPanel Example](../../media/development/core/ui-component-framework/stats-panel-example.png){ .on-glb }
</div>
<!-- markdownlint-enable no-inline-html -->

> ### Text Panels Family

Reviewer: ```suggestion
### BaseTextPanel
```

`ObjectTextPanel` renders content from a specific field of an object in the context. It simplifies the display of object attributes in various text formats (Markdown, JSON, YAML, etc.).

[Code reference](../../code-reference/nautobot/apps/ui.md#nautobot.apps.ui.ObjectTextPanel)

#### ObjectTextPanel Examples

```python
from nautobot.apps.ui import ObjectTextPanel, SectionChoices

ObjectTextPanel(
    object_field="description",
    render_as=ObjectTextPanel.RenderOptions.MARKDOWN,
    render_placeholder=True,
    label="Description",
    section=SectionChoices.FULL_WIDTH,
    weight=500
)
```

### 6. TextPanel

`TextPanel` renders content from a specified context field. It provides a simple way to display text content in various formats (Markdown, JSON, YAML, plaintext, or code) from the rendering context.

[Code reference](../../code-reference/nautobot/apps/ui.md#nautobot.apps.ui.TextPanel)

#### TextPanel Examples

```python
from nautobot.apps.ui import TextPanel, SectionChoices

TextPanel(
    context_field="text",
    render_as=TextPanel.RenderOptions.CODE,
    render_placeholder=True,
    label="Custom Content",
    section=SectionChoices.FULL_WIDTH,
    weight=600
)
```

> ### Table Panels Family
> 
> ---

Reviewer: ```suggestion
```

# Nautobot UI Framework

## Table of Contents
- [Introduction](#introduction)
- [Getting Started](#getting-started)
- [Core Concepts](#core-concepts)
- [Panel Types](#panel-types)
- [Complete Example](#complete-example)
- [Layouts and Sections](#layouts-and-sections)
- [Best Practices](#best-practices)
- [Troubleshooting](#troubleshooting)

## Introduction

The Nautobot UI Framework revolutionizes how you create object detail views in your applications. Instead of writing HTML templates, you declaratively define your UI structure using Python objects, resulting in consistent, maintainable, and responsive interfaces.

//TODO: (include image) <img src="/api/placeholder/800/400" alt="UI Framework Example" />
*Example of a detail view created with the UI Framework*

### Why Use the UI Framework?
- ✨ **Reduced Development Time**: Eliminate boilerplate HTML/CSS
- 🎯 **Consistency**: Automatic adherence to Nautobot's design patterns

> - 📱 **Responsive by Default**: Built-in mobile-friendly layouts

Reviewer: Remove

- 🔄 **Reusable Components**: Compose views from pre-built panels
- 🛠 **Extensible**: Easy to customize and extend

## Getting Started

### Basic Setup

> 1. Create a view that inherits from `NautobotViewSetMixin` or `ObjectView`, in this example `ObjectView`:

Reviewer: All new UI development should be using `NautobotUIViewSet`.

```python
from nautobot.core.views.generic import ObjectView

class ExampleUIViewSet(ObjectView):
    queryset = Example.objects.all()
    ...
    
    object_detail_content = ObjectDetailContent(
        panels=[
            ObjectFieldsPanel(
                section=SectionChoices.LEFT_HALF,
                weight=100,
                fields="__all__"
            )
        ]
    )
```

//TODO: (include image) <img src="/api/placeholder/600/300" alt="Basic Panel Layout" />
*Basic panel layout with ObjectFieldsPanel*

## Core Concepts

### ViewSet Configuration
The UI Framework is built around the concept of configurable ViewSets. Each ViewSet defines:
- Query handling
- Panel layouts
- Content organization

### Panels
Panels are the building blocks of your UI. They contain specific types of content and can be positioned within sections.

//TODO: (include image) <img src="/api/placeholder/700/350" alt="Panel Structure" />
*Panel structure showing different sections and layouts*

## Panel Types

### Base Panel Properties
The Panel component serves as a base class for creating individual display panels within a Layout system.

#### Configuration Parameters

> | Parameter | Required | Default | Description |

Reviewer: This table can/should be replaced with a link to the autogenerated `Panel` documentation (in `code-reference/nautobot/apps/ui/`) rather than duplicating information.

|-----------|----------|---------|-------------|
| `label` | No | `""` | The display label for the panel. If empty, no label will be shown. |

> | `section` | No | `SectionChoices.FULL_WIDTH` | Determines which layout section the panel belongs to. |

Reviewer 2: I wonder if it's worth to add all possible values here? Or just add a link to the `SectionChoices` documentation? The same goes to the `RenderOptions`.

Reviewer 2: Ok, I've found documentation of enums. So maybe just add a link here?

> | `weight` | Yes | - | Integer determining component's rendering order relative to peers |

Reviewer 2: I would add one more statement about fact that full width panels will be under half's and that panels with smaller weights will be higher.
